### ISSUE TITLE
Documented rotation method doesn't work with ATLAS as the BLAS implementation

### ISSUE THREAD

**Issue by capnrefsmmat** [r-spatial/sf#2067]
**Describe the bug**

The rotation method described in the [sf3 vignette](https://r-spatial.github.io/sf/articles/sf3.html) produces invalid geometries on platforms using [ATLAS](https://math-atlas.sourceforge.net/) as their BLAS implementation. This includes [a CRAN additional check platform](https://www.stats.ox.ac.uk/pub/bdr/Rblas/README.txt). Multiplying geometries by a rotation matrix works on most platforms, but with ATLAS, it results in evaluation errors.

I'm not sure whose problem this is: should ATLAS not be giving different results? should sf be doing something differently to the geometries? But regardless, given the rotation method is in your documentation, I figured you'd want to know.

We discovered this because our package [failed the CRAN additional check](https://cran.r-project.org/web/checks/check_results_covidcast.html), and we'll have to work around the issue to remain on CRAN. I expect we'll simply work around to avoid the rotation, since that's easy in our case.

**To Reproduce**

Using the latest stable sf and the sample from the vignette, with ATLAS we get:

```
> library(sf)
Linking to GEOS 3.10.2, GDAL 3.4.3, PROJ 8.2.1; sf_use_s2() is TRUE
> nc = st_read(system.file("shape/nc.shp", package="sf"), quiet=TRUE)
> ncg = st_geometry(nc)
> cntrd = st_centroid(ncg)
> rot = function(a) matrix(c(cos(a), sin(a), -sin(a), cos(a)), 2, 2)
> st_is_valid((ncg - cntrd) * rot(pi/4) * .75)
  [1] TRUE   NA TRUE   NA   NA TRUE TRUE   NA TRUE   NA   NA   NA   NA   NA TRUE
 [16] TRUE TRUE TRUE   NA TRUE TRUE TRUE TRUE TRUE TRUE   NA   NA TRUE   NA TRUE
 [31] TRUE TRUE TRUE TRUE TRUE   NA   NA TRUE TRUE   NA TRUE   NA   NA TRUE TRUE
 [46]   NA   NA TRUE   NA   NA TRUE   NA TRUE   NA TRUE   NA   NA TRUE   NA TRUE
 [61]   NA   NA   NA   NA TRUE TRUE   NA   NA TRUE TRUE TRUE TRUE TRUE   NA   NA
 [76]   NA TRUE TRUE TRUE TRUE   NA TRUE TRUE   NA   NA   NA TRUE   NA TRUE TRUE
 [91] TRUE TRUE TRUE   NA TRUE   NA   NA TRUE   NA TRUE
> st_make_valid((ncg - cntrd) * rot(pi/4) * .75)
Error in CPL_geos_make_valid(x, geos_method, geos_keep_collapsed) : 
  Evaluation error: IllegalArgumentException: Points of LinearRing do not form a closed linestring.
```

I also checked `st_is_valid(ncg)`, `st_is_valid(cntrd)`, and `st_is_valid(ncg - cntrd)`; the `NA`s only appear once there's a rotation.
 
On macOS (GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1), however, `st_is_valid()` returns `TRUE` for all entries, and `st_make_valid()` does not raise any error. Based on our package's CRAN check results, the only platform this occurs on is their ATLAS platform.

In our package's case, this only occurs when rotating specific geometries, and it seems inconsistent between machines which specific geometries fail -- but because we're making plots of counties, states, and MSAs in the United States, there are lots of geometries.

**Additional context**

Our package issue is here: https://github.com/cmu-delphi/covidcast/issues/605

We have a Dockerfile for setting up an image with ATLAS, and this is what I used to reproduce the problem: https://github.com/cmu-delphi/covidcast/issues/605#issuecomment-1350345564

<details>
This is from our Docker image with ATLAS:

```
> sessionInfo()
R Under development (unstable) (2022-12-19 r83479)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Fedora Linux 36 (Container Image)

Matrix products: default
BLAS/LAPACK: /usr/lib64/atlas/libsatlas.so.3.10;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sf_1.0-9        vdiffr_1.0.4    curl_4.3.3      dplyr_1.0.10   
 [5] MMWRweek_0.1.3  mockery_0.4.3   httptest_4.1.0  tibble_3.1.8   
 [9] covidcast_0.4.4 testthat_3.1.6 

loaded via a namespace (and not attached):
 [1] gtable_0.3.1       ggplot2_3.4.0      htmlwidgets_1.6.0  devtools_2.4.5    
 [5] remotes_2.4.2      processx_3.8.0     callr_3.7.3        vctrs_0.5.1       
 [9] tools_4.3.0        ps_1.7.2           generics_0.1.3     proxy_0.4-27      
[13] fansi_1.0.3        pkgconfig_2.0.3    KernSmooth_2.23-20 desc_1.4.2        
[17] lifecycle_1.0.3    farver_2.1.1       compiler_4.3.0     stringr_1.5.0     
[21] brio_1.1.3         munsell_0.5.0      httpuv_1.6.7       htmltools_0.5.4   
[25] usethis_2.1.6      class_7.3-20.1     tidyr_1.2.1        later_1.3.0       
[29] pillar_1.8.1       crayon_1.5.2       urlchecker_1.0.1   ellipsis_0.3.2    
[33] classInt_0.4-8     cachem_1.0.6       wk_0.7.1           sessioninfo_1.2.2 
[37] mime_0.12          tidyselect_1.2.0   digest_0.6.31      stringi_1.7.8     
[41] purrr_0.3.5        labeling_0.4.2     rprojroot_2.0.3    fastmap_1.1.0     
[45] grid_4.3.0         colorspace_2.0-3   cli_3.4.1          magrittr_2.0.3    
[49] pkgbuild_1.4.0     utf8_1.2.2         e1071_1.7-12       withr_2.5.0       
[53] waldo_0.4.0        prettyunits_1.1.1  scales_1.2.1       promises_1.2.0.1  
[57] httr_1.4.4         memoise_2.0.1      shiny_1.7.4        miniUI_0.1.1.1    
[61] s2_1.1.1           profvis_0.3.7      rlang_1.0.6        Rcpp_1.0.9        
[65] xtable_1.8-4       glue_1.6.2         DBI_1.1.3          pkgload_1.3.2     
[69] jsonlite_1.8.4     rstudioapi_0.14    R6_2.5.1           fs_1.5.2          
[73] units_0.8-1       
> sf::sf_extSoftVersion()
          GEOS           GDAL         proj.4 GDAL_with_GEOS     USE_PROJ_H 
      "3.10.2"        "3.4.3"        "8.2.1"         "true"         "true" 
          PROJ 
       "8.2.1" 
```
</details>

**Comment by edzer** [comment]
Great report, I could reproduce using the docker image. That rotation obviously goes into a matrix product computation, [here](https://github.com/r-spatial/sf/blob/main/R/arith.R#L101), which will call atlas routines. Although the input geometry has closed rings (first and last points, rows in the matrix, are identical), after rotating with atlas this is no longer true: for the second geometry I find
```r
> ncg2 = ncg[2]
> cntrd2 = cntrd[2]
> x = (ncg2 - cntrd2) * rot(pi/4) * .75
> p1 = x[[1]][[1]][[1]][1,] # first
> p26 = x[[1]][[1]][[1]][26,] # last
> p1
[1] -0.127547393 -0.005824681
> p26
[1] -0.127547393 -0.005824681
> p1 - p26
[1]  0.000000e+00 -4.336809e-18
> print(p1 - p26, digits = 20)
[1]  0.000000000000000000e+00 -4.336808689942017736e-18
```
So with identical matrix rows, atlas manages to do a matrix product with the same matrix that results in different rows - this is what was meant by `Points of LinearRing do not form a closed linestring`. How about that. We can of course "fix" this in `sf` by postprocessing all geometries after a `%*%`, possibly conditioning on
```r
> grepl("atlas", sessionInfo()$BLAS)
[1] TRUE
```
but I'd be interested in what changed in atlas that now causes this, and whether this was a good change.

**Comment by edzer** [comment]
Full R script:
```r
library(sf)
nc = st_read(system.file("shape/nc.shp", package="sf"), quiet=TRUE)
ncg = st_geometry(nc)
cntrd = st_centroid(ncg)
rot = function(a) matrix(c(cos(a), sin(a), -sin(a), cos(a)), 2, 2)
st_is_valid((ncg - cntrd) * rot(pi/4) * .75)
ncg2 = ncg[2]
cntrd2 = cntrd[2]
x = (ncg2 - cntrd2) * rot(pi/4) * .75
x[[1]] # shows the full ring:
p1 = x[[1]][[1]][[1]][1,]
p26 = x[[1]][[1]][[1]][26,]
p1
p26
p1 - p26
print(p1 - p26, digits = 20)
q1 = ncg2[[1]][[1]][[1]][1,]
q26 = ncg2[[1]][[1]][[1]][26,]
print(q1 - q26, digits = 20)
```

**Comment by LHMarshall** [comment]
I just wondered if this issue has been taken any further... has it been reported to the ATLAS project? My package uses the rotation a lot and has unfortunately been bumped off CRAN due to the issue. I'm just debating my options. Thanks

**Comment by edzer** [comment]
I didn't, I wouldn't know where to start. If ATLAS keeps doing this we could think about preprocessing polygons before they go into GEOS (or s2) for validity checking to correct for such small shifts in end point coordinates.

**Comment by edzer** [comment]
The problem, b.t.w., is also solved by setting a precision, e.g.
```r
> st_make_valid(x)
Error in CPL_geos_make_valid(x, geos_method, geos_keep_collapsed) : 
  Evaluation error: IllegalArgumentException: Points of LinearRing do not form a closed linestring.
> st_make_valid(st_set_precision(x,1e8))
Geometry set for 1 feature 
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: -0.1275474 ymin: -0.1387147 xmax: 0.156867 ymax: 0.1601199
CRS:           NA
POLYGON ((-0.1275474 -0.00582468, -0.1205193 0....
> st_is_valid(st_set_precision(x,1e8))
[1] TRUE
```
Using value 1e8, as example, rounds to 8 digits.

**Comment by rsbivand** [comment]
You mention macOS as a contrast. Which processor architecture and which version of ATLAS are you using? Note that R in general advises care in trying to use optimised BLAS. If intel, some BLAS do not use enhanced double precision IIRC.

**Comment by capnrefsmmat** [comment]
I was unclear. On macOS I'm using whichever BLAS R uses by default, not ATLAS. (We had a hard enough time getting ATLAS working on Linux, to match the CRAN setup, that we didn't also run it on macOS.) This is an Intel machine.

We did have it apparently work on Linux with ATLAS, though. We tried both [using flexiblas and compiling R with ATLAS](https://github.com/cmu-delphi/covidcast/issues/605#issuecomment-1349005546) and our tests passed; when we switched to the approach used on CRAN's system, which involves symlinking R's BLAS to ATLAS, the tests failed.

That suggests the issue depends on how ATLAS is configured. If there's any information I can report from those sessions that would be helpful, I can try to extract it.
